Credits show contributors name if available

HanLap opened the pull request:
Closes  #456

This changes the update credits workflow step.

Contributors are listed as follows:
- if name exists: `<name> (<login>)`
- otherwise: `<login>`

captaincolonelfox reviewed (dismissed):
Awesome work! Thank you for your contribution

HanLap commented on .github/workflows/update-credits.js:
```diff
@@ -13,11 +13,26 @@ module.exports = async (github, context) => {
+
+          const { name, login } = res.data;
+
+          if (!name) {
+            return `${login}\n`;
+          }
+
+          return `${name} (${login})\n`;
```
Would you respect casing in this case? 

E.g should login `hanlap` and name  `HanLap` result in `hanlap` or in `HanLap (hanlap)`?

captaincolonelfox replied:
Hmm, it's a good question. If you setup a different casing in `login` and `name`, so the casing might be important for you. I guess, we need to prioritize `name` if values logically the same, but casing is different
Like that
```suggestion
          if (!name) {
            return `${login}\n`;
          }
          if (name.toLowerCase() === login.toLowerCase()) {
            return `${name}\n`;
          }
          return `${name} (${login})\n`;
```

HanLap commented:
I just realized that my implementation is straining the API rate limit quite a bit. An API call is required for each Contributor. That currently means 93 calls. 

If this project is subject to the same rate limit as personal users (1000/h), it is very likely to hit this limit.

Potential solutions would be:
- only build the detailed credit list for a main branch release.
- somehow cache the credits between workflows.

captaincolonelfox commented:
@HanLap 
> API rate limit

Nice catch, I wasn't aware of that. We can pass `headers: {"If-Modified-Since": date}` to `github.repos.listContributors` request so it will return `304 Not Modified` status code if there is no new contributors and no need to rebuild `credits.txt`

There is a way to cache/restore file ([actions/cache](https://github.com/actions/cache)) and a way to delete used cache ([github.actions.deleteActionsCacheById](https://docs.github.com/en/rest/actions/cache#delete-github-actions-caches-for-a-repository-using-a-cache-key))

It just require a further investigation how to properly use it

idbrii commented:
I just read about this project from lazerwalker's [Managing a game dev community with GitHub Actions](https://github.blog/2022-07-06-managing-a-game-dev-community/) post. She mentioned crediting in the post:

> An important note on inclusivity: although both git commit history and the GitHub API contain contributor information, we very consciously chose to use the [GitHub contributors API](https://docs.github.com/en/rest/repos#list-contributors) to use GitHub usernames rather than git commit names. If someone wants to change their name in the credits (for example, they’ve transitioned, and previous contributions were made under their deadname), updating a git-based credits screen would require rewriting git history, which can be disruptive for an open source project. Conversely, pulling real-time data from the GitHub API helped us ensure that we were always referring to people how they wanted to be.

I guess this PR still aligns with that view because it gets the name list every time a build is created? And caching would be time-limited so eventually all names get re-fetched?

Regardless, cool project!

captaincolonelfox commented:
@idbrii
Oh yeah, we need to take account of that aswell. Not sure if `actions/cache` have something built-in, but we can have week-number in cache key name. So even if number of contributors didn't changed, the credits will be rebuild anyway - so new names will be fetched

HanLap commented:
Okay, here is what i implemented:
Credits are cached for a single day. As long as there is a cached credits item for the given day the update credits step won't run at all. 
Furthermore, because `actions/cache` will only search for cache items of the current branch and its parent branch, I added a daily workflow that will build the credits cache for the main branch.

captaincolonelfox commented on .github/workflows/daily-credits.yaml:
```diff
@@ -0,0 +1,34 @@
+            await script(github, context)
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+  
```
We can add this steps to `daily credits build` workflow to push a new build as well

Though it maybe bad idea to push a new build everyday even if number of contributors or their usernames didn't change. So we might want to have some sort of hash check of `credits.txt` for a previous day - and if file didn't change we won't run next steps

```suggestion
       - name: export with Godot Engine
        uses: firebelley/godot-export@v3.0.0
        with:
          create_release: false
          archive_export_output: true
          godot_executable_download_url: https://downloads.tuxfamily.org/godotengine/3.4.4/Godot_v3.4.4-stable_linux_headless.64.zip
          godot_export_templates_download_url: https://downloads.tuxfamily.org/godotengine/3.4.4/Godot_v3.4.4-stable_export_templates.tpz
          relative_export_path: ./
          relative_project_path: ./
      - name: deploy to itch.io
        uses: josephbmanley/butler-publish-itchio-action@master
        env:
          BUTLER_CREDENTIALS: ${{ secrets.ITCH_API_KEY }}
          CHANNEL: html
          ITCH_GAME: a-little-game-called-mario
          ITCH_USER: iznaut
          PACKAGE: HTML5.zip
          VERSION: credits-update
```

captaincolonelfox replied:
We can have few outputs in `date` step 
```suggestion
  - name: Get current date
    id: date
    run: |
       echo ::set-output name=date::"$(date +'%Y-%m-%d')"
       echo ::set-output name=yesterday::"$(date --date='yesterday' +'%Y-%m-%d')"
```

And have additional step to get credits for previous date cache
```suggestion
      - name: yesterday credits
        id: yesterday-credits
        uses: actions/cache@v3.0.0
        with:
          key: credits-${{ steps.date.outputs.yesterday }}
          path: credits-yesterday.txt
```

Then we can check for their hash sums in our build steps
```suggestion
if: ${{ hashFiles('credits.txt') != hashFiles('credits-yesterday.txt') }}
```

But this might fail if `credits-yesterday.txt` is not exist, so we want another additional step
```suggestion
-  name: stub yesterday credits
   id: stub-yesterday-credits
   if: ${{ steps.yesterday-credits.outputs.cache-hit != 'true' }}
   run: touch credits-yesterday.txt
```

HanLap replied:
It does not seem possible to restore a file from a cache item using a different path
```suggestion
        - name: yesterday credits
        id: yesterday-credits
        uses: actions/cache@v3.0.0
        with:
          key: credits-${{ steps.date.outputs.yesterday }}
          path: credits-yesterday.txt
```
This action will search for a file called `credits-yesterday.txt` in the corresponding cache item.

Sadly i cannot think of an alternative way to solve this problem.

captaincolonelfox replied:
@HanLap
Oh so you want to say, that both `key` and `path` used to determine what file was cached?
Yeah, I guess I I misunderstood that and I was thinking that `key` - is determine the file and a `path` just used to save/restore file for that `key`

So in that case, we can restore `yesterday-credits.txt` first, then rename it, and only then restore `credits.txt`:
```
- id: date
   run: |
      echo ::set-output name=date::"$(date +'%Y-%m-%d')"
      echo ::set-output name=yesterday::"$(date --date='yesterday' +'%Y-%m-%d')"

- id: restore-yesterday-credits
  uses: actions/cache@v3.0.0
  with:
    key: credits-${{ steps.date.outputs.yesterday }}
    path: credits.txt
      
- id: stub-yesterday-credits
   if: ${{ steps.restore-yesterday-credits.outputs.cache-hit != 'true' }}
   run: touch credits.txt
     
- run: mv credits.txt credits-yesterday.txt

- id: cache-credits
  uses: actions/cache@v3.0.0
  with:
    key: credits-${{ steps.date.outputs.date }}
    path: credits.txt
    
- name: update credits
   uses: actions/github-script@v3
   with:
     script: |
       const path = require('path');
       const script = require(path.resolve('.github/workflows/update-credits.js'));
       await script(github, context)
       
- if: ${{ hashFiles('credits.txt') != hashFiles('credits-yesterday.txt') }}
  #BUILD
```

hughesjs commented:
Is it worth noting that some people might not want their username being auto-translated into their real name even if it's available on their GitHub account?

A lot of people would rather build a reputation based on their pseudonym than their real name.

Maybe add a whitelist?

captaincolonelfox commented:
@hughesjs This script will use both `<name> (login)`. But maybe we need to consider highlighting `login` instead of `name` and switch them. 
Creating some `login-only-filter.txt` file is might be a good idea for those who only wants to have their `login` in the credits. But I would like to have it as a separate pull request after this one get merged, because it's already getting over complicated

lazerwalker commented:
To take a step back, I believe this PR does three things:

1. Consistently shows both the GH username and display name in the credits
2. If I understand correctly, this changes it so credits are updated at most once a day. If I'm the second PR merged in a day, my name won't be added to the credits until the next day
3. Guarantees that the credits are updated even on days when no new contributions are merged

Under the current system / without this code, because GH inconsistently caches credit data, sometimes (but not always) someone whose code gets merged in might not see their name in the credits until a subsequent PR, which could be days. So (3) fixes an existing issue with credits not showing up, in addition to the new functionality added in (1).

As it stands, even if we say that (1) is unequivocally a positive change (which I'm not sure is the case), (2) feels like a large step back. I'm not sure the complexity we're adding is necessarily worth it here? Running a daily "do we have credits that need to be updated?" check makes a lot of sense to me to capture that edge case (and is something that could happen independent of the rest of this logic), but I'm not sure the questionable benefit of showing display names outweighs the credit lag being applied to a bunch of new contributions plus a bunch more workflow code to maintain. 

Curious to hear others' thoughts!

hughesjs commented:
Just a thought... Could we not do this in the hooks rather than as an action?

captaincolonelfox commented:
@lazerwalker @HanLap 
What do you think about this solution? It would rebuild credits and cache them everyday (and push new build if credits are different from previous day), as well as rebuilding them on new contributor within a day and “update” cache - so every other build still using latest cached credits

@HanLap If it’s okay with you want, I can implement that change and push commit to your branch

HanLap commented:
I am sorry for not responding until now. \~_\~ 

@captaincolonelfox It would be nice if you could implement the change.

captaincolonelfox commented:
@HanLap 
I'm sorry, I tried to implement that, but immediately stuck

Didn't test those enough:
- `{"If-Modified-Since": date}` includes change in number of commits for any contributor - so it basically was triggered on any commit, not only on new contributor
- `"author_association": "NONE"` in `context` was only available on `pull_request` events, but not on main branch `push` event. And while we can rebuild credits on pull requests from new contributors, that cache will not be used in main branch later on

We need somehow check if push was done by `first-time contributor`, but I don't have anything particular in mind
